### PR TITLE
[refactor](array-type) optimize the array_distinct and array_sort fun…

### LINKED ISSUE #10052: [Feature] Add lots of Array Functions
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/incubator-doris/issues?q=is%3Aissue) and found no similar issues.

### Description

We have supported `array` data type recently, all pr(s) about `array` please refer to https://github.com/apache/incubator-doris/issues/7570 .

Now we need to add lots of array functions to make `array` more powerful.
If anybody is interesting to `array`, please let me know.

Array Functions DONE:
- size/cardinality
- array_position
- array_contains
- element_at

Array Functions DOING:
- array_sum
- array_avg
- array_min
- array_max
- array_product
- array_distinct
- array_remove
- arrays_overlap

Array Functions TODO:
- array_union
- array_intersect
- array_except
- reverse
- concat_ws
- array_join
- array_sort
- array_slice

### Use case

_No response_

### Related issues

https://github.com/apache/incubator-doris/issues/7570

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by carlvinhust2012**
# Proposed changes
1. this pr is used to optimize the array_distinct and array_sort function.

Issue Number: close #10052

## Checklist(Required)

1. Type of your changes:
    - [ ] Improvement
    - [ ] Fix
    - [ ] Feature-WIP
    - [ ] Feature
    - [ ] Doc
    - [x] Refator
    - [ ] Others: 
3. Does it affect the original behavior: 
    - [ ] Yes
    - [x] No
    - [ ] I don't know
4. Has unit tests been added:
    - [ ] Yes
    - [ ] No
    - [x] No Need
5. Has document been added or modified:
    - [ ] Yes
    - [ ] No
    - [x] No Need
6. Does it need to update dependencies:
    - [ ] Yes
    - [x] No
7. Are there any changes that cannot be rolled back:
    - [ ] Yes
    - [x] No

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: be/src/vec/functions/array/function_array_sort.cpp [hunk 1]
```diff
@@ -15,12 +15,203 @@
...
+                    (*dest_null_map).push_back(false);
+                }
+            }
+            dest_offsets.push_back(curr_src_offset);
+            prev_src_offset = curr_src_offset;
+        }
+        return true;
+    }
+
+    static bool _execute_by_type(const IColumn& src_column, const ColumnArray::Offsets& src_offsets,
```
Comment (xy720) [inline be/src/vec/functions/array/function_array_sort.cpp:166]: This function is completely duplicate with the one in function_array_distinct.cpp.
You need to think for an abstract method or impl.

File: be/src/vec/functions/array/function_array_sort.cpp [hunk 2]
```diff
@@ -15,12 +15,203 @@
...
+            prev_src_offset = curr_src_offset;
+        }
+
+        return true;
+    }
+
+    static bool _execute_string(const IColumn& src_column, const ColumnArray::Offsets& src_offsets,
+                                IColumn& dest_column, ColumnArray::Offsets& dest_offsets,
+                                const NullMapType* src_null_map, NullMapType* dest_null_map) {
+        const ColumnString* src_data_concrete = reinterpret_cast<const ColumnString*>(&src_column);
```
Comment (xy720) [inline be/src/vec/functions/array/function_array_sort.cpp:109]: Duplicate with function_array_distinct.cpp
If it is hard to add an abstract method, you can put the duplicate code into function_array_utils.h and use a common util method instead.

File: be/src/vec/functions/array/function_array_sort.cpp [hunk 3]
```diff
@@ -15,12 +15,203 @@
...
+        if (!src_data_concrete) {
+            return false;
+        }
+
+        ColumnString& dest_column_string = reinterpret_cast<ColumnString&>(dest_column);
```
Comment (xy720) [inline be/src/vec/functions/array/function_array_sort.cpp:114]: Duplicate too.

**Review by xy720 — COMMENTED** [review]
There are still much duplicate code between function_array_sort.cpp and function_array_distinct.cpp. Is there any good idea to extract these code out?

**Comment by carlvinhust2012** [comment]
> There are still much duplicate code between function_array_sort.cpp and function_array_distinct.cpp. Is there any good idea to extract these code out?

ok，I will re-check again.

**Comment by carlvinhust2012** [comment]
reopen this pr.

File: be/src/vec/functions/array/function_array_sort.cpp (hunk 1 above)
Comment (carlvinhust2012) [reply be/src/vec/functions/array/function_array_sort.cpp]: done

**Pull request closed without merging**